### DISCUSSION TITLE
Does netdata apps.plugin support dynamic grouping by `/proc/pid/cmdline` ?

### DISCUSSION THREAD

**Discussion by xxm404** [netdata/netdata#15285] [Q&A]
For example when run some script like this command
```
/usr/local/php72/bin/php -c /home/work/env/php/etc/php.ini run_test0.php
/usr/local/php72/bin/php -c /home/work/env/php/etc/php.ini run_test1.php
/usr/local/php72/bin/php -c /home/work/env/php/etc/php.ini run_test2.php
...
```
Whether netdata can be grouped according to the content of the corresponding process `/proc/pid/cmdline` (hope this can be dynamic, and will be automatically recognized when new scripts are added)

**Comment by ilyam8** [comment]
Yes, it does support it.

**Reply by xxm404** [reply]
Thanks for the quick reply
In my example
```
/usr/local/php72/bin/php -c /home/work/env/php/etc/php.ini run_test0.php
/usr/local/php72/bin/php -c /home/work/env/php/etc/php.ini run_test1.php
/usr/local/php72/bin/php -c /home/work/env/php/etc/php.ini run_test2.php
...
```
i want get three groups (run_test0, run_test1, run_test2) at netdata application section, when starting a new script (/usr/local/php72/bin/php -c /home/work/env/php/etc/php.ini run_test3.php), I hope to dynamically add a new metric group (run_test3), how can i modify the configuration of the apps.plugin.

**Reply by ilyam8** [reply]
Add before [this line](https://github.com/netdata/netdata/blob/435cf76e736469ec02649b19095b225586f42f61/collectors/apps.plugin/apps_groups.conf#L108)

```
php-test0: *run_test0*
php-test1: *run_test1*
php-test2: *run_test2*
```

**Reply by xxm404** [reply]
This way it seems that every time I start a new task, I need to modify the configuration file of netdata and restart it.
I was wondering if it is possible to implement logic like the following, Suppose the cmdline corresponding to my task has three fields( cat /proc/pid/cmdline | strings)
```
cat /proc/17322/cmdline | strings 
/usr/local/php72/bin/php # field0
/home/work/env/php/etc/php.ini # field1
run_test.php # field2
```
Here I express my thoughts in a syntax similar to sql
```
select cpu,mem,process,network from netdata where field0 = '/usr/local/php72/bin/php' group by field2
```
Use the first field to get a set of indicators, and use the third field for group display

**Reply by ilyam8** [reply]
If you want to dynamically create groups (chart dimensions) - no, this is not supported.

**Reply by andrewm4894** [reply]
i guess you might need a little script to be able to append the new groups to apps_groups.conf and then restart netdata each time you kick off one of these new processes that you want to have its own dimension in netdata.

@ilyam8 am curious would the netdatacli ever be able to support something like this or is it always going to need a netdata restart for the config change to apps_groups.conf?

**Reply by xxm404** [reply]
> If you want to dynamically create groups (chart dimensions) - no, this is not supported.

Thanks for the reply.

**Reply by xxm404** [reply]
> i guess you might need a little script to be able to append the new groups to apps_groups.conf and then restart netdata each time you kick off one of these new processes that you want to have its own dimension in netdata.

Maybe I can try to customize the apps.plugin to suit my needs, but then I won't be able to keep updating the netdata agent.

**Reply by Ferroin** [reply]
I can see potential value for other users in having a feature like this. If you end up customizing things to support it and can make the customizations generic enough that they’re not specific to your setup, then I think it’s pretty likely that a PR with those changes would at minimum get reviewed, and quite possibly merged as well, at which point you wouldn’t need to worry about keeping a customized version set up locally.

**Reply by andrewm4894** [reply]
I wonder how this might work.

Maybe you would need to have a specific sort of "watch" pattern, and all stuff under that would get broken out.

So in your case the "watch pattern" could be "run_test*.php" such that all processes matching that pattern would get split out into their own "app". 

I think it's a good idea but we'd need to sort of define some sensible way to go about it I think. 

Not 100% sure myself but just throwing an idea out there.

**Reply by andrewm4894** [reply]
Maybe it's just a case of adding some decoration to each group to say to consolidate them all into to one dim (current default) or just split them into individual dims (some new capability).

**Reply by xxm404** [reply]
> I can see potential value for other users in having a feature like this. If you end up customizing things to support it and can make the customizations generic enough that they’re not specific to your setup, then I think it’s pretty likely that a PR with those changes would at minimum get reviewed, and quite possibly merged as well, at which point you wouldn’t need to worry about keeping a customized version set up locally.  

I will first examine the implementation of the NETDATA APPS.plugin, incorporate the necessary functionality to meet my specific requirements, and subsequently strive to enhance its generality (although I cannot ascertain if this functionality will be required by other users).

**Reply by xxm404** [reply]
> Maybe it's just a case of adding some decoration to each group to say to consolidate them all into to one dim (current default) or just split them into individual dims (some new capability).

Sounds good, let me describe my use case, I want to monitor a set of offline tasks (similar to scheduled tasks, etc.), the common denominator of this set of tasks is the use of similar startup commands (/usr/local/php72/bin/php -c /home/work/env/php/etc/php.ini xxx .php) I want to see the metrics for each task, so it would be ideal if you could support fuzzy matching and grouping by some rules, of course, the number of tasks deployed per machine should be controlled, otherwise performance may be affected here
